include/ginkgo/core/distributed/dd_matrix.hpp

```diff
@@ -0,0 +1,400 @@
+
+public:
+    using value_type = ValueType;
+    using index_type = GlobalIndexType;
+    using local_index_type = LocalIndexType;
+    using global_index_type = GlobalIndexType;
+    using global_matrix_type =
+        Matrix<ValueType, LocalIndexType, GlobalIndexType>;
+    using global_vector_type =
+        gko::experimental::distributed::Vector<ValueType>;
+    using local_vector_type = typename global_vector_type::local_vector_type;
+
+    using EnableLinOp<DdMatrix>::convert_to;
+    using EnableLinOp<DdMatrix>::move_to;
+    using ConvertibleTo<DdMatrix<next_precision_base<ValueType>, LocalIndexType,
+                                 GlobalIndexType>>::convert_to;
+    using ConvertibleTo<DdMatrix<next_precision_base<ValueType>, LocalIndexType,
+                                 GlobalIndexType>>::move_to;
+
+    void convert_to(DdMatrix<next_precision_base<value_type>, local_index_type,
+                             global_index_type>* result) const override;
+
+    void move_to(DdMatrix<next_precision_base<value_type>, local_index_type,
+                          global_index_type>* result) override;
+
+    /**
+     * Reads a square matrix from the device_matrix_data structure and a global
+     * partition.
+     *
+     * The global size of the final matrix is inferred from the size of the
+     * partition. Both the number of rows and columns of the device_matrix_data
+     * are ignored.
+     *
+     * @note The matrix data can contain entries for rows other than those owned
+     *        by the process. The local matrix still considers these and the
+     *        restriction and prolongation operators take care of fetching /
+     *        re-distributing the corresponding vector entries.
+     *
+     * @param data  The device_matrix_data structure.
+     * @param partition  The global partition.
+     */
+    void read_distributed(
+        const device_matrix_data<value_type, global_index_type>& data,
+        std::shared_ptr<const Partition<local_index_type, global_index_type>>
+            partition);
+
+    /**
+     * Reads a square matrix from the matrix_data structure and a global
+     * partition.
+     *
+     * @see read_distributed
+     *
+     * @note For efficiency it is advised to use the device_matrix_data
+     * overload.
+     */
+    void read_distributed(
+        const matrix_data<value_type, global_index_type>& data,
+        std::shared_ptr<const Partition<local_index_type, global_index_type>>
+            partition);
+
+    /**
+     * Get read access to the stored local matrix.
+     *
+     * @return  Shared pointer to the stored local matrix
+     */
+    std::shared_ptr<const LinOp> get_local_matrix() const { return local_mtx_; }
+
+    /**
+     * Get read access to the stored restriction operator.
+     *
+     * @return  Shared pointer to the stored restriction operator.
+     */
+    std::shared_ptr<const global_matrix_type> get_restriction() const
```

Comment:
as mentioned before, maybe this should return a `gko::LinOp` to allow us more flexibility when changing the internal data structure.